### PR TITLE
Testing:upload/download and xrd integration tests #2311

### LINKED ISSUE #2311: Functional download/upload tests with xrootd containers
Motivation
----------
At the moment we only have very limited upload/download tests either using mock calls or direct posix protocols. One possibility to expand these to actualy functional tests would be to use xrootd docker containers and start up two xrootd storages in the test environment. This would allow us to test upload/download also with different protocols, such as gfal, which would also help enormously to validate gfal for python 3.6.

### PR DESCRIPTION
**Pull request by maany**
Functional download/upload tests with xrootd containers
------------------
- sample run: 
  - failure: https://github.com/maany/rucio/runs/1459016452?check_suite_focus=true
  - success: https://github.com/maany/rucio/runs/1459042130?check_suite_focus=true
- test_rucio_server.py and test_rse_protocol_xrootd,py check for xrd[1-3] containers and perform (upload, download), (upload, delete, rename, exists) tests
- new GitHub Actions workflow with docker-compose-storage.yaml. The tests can be run locally using the exact steps described in the integration-tests.yml workflow file

### CODE REVIEW

File: lib/rucio/tests/test_rse_protocol_xrootd.py [hunk 1]
```diff
@@ -173,7 +201,7 @@ def test_put_mgr_FileReplicaAlreadyExists_single(self):
         with pytest.raises(exception.FileReplicaAlreadyExists):
             self.mtc.test_put_mgr_FileReplicaAlreadyExists_single()
 
-    # MGR-Tests: DELETE
+    # # MGR-Tests: DELETE
```
Comment (maany) [inline lib/rucio/tests/test_rse_protocol_xrootd.py]: redundant change

File: lib/rucio/tests/test_rse_protocol_xrootd.py [hunk 2]
```diff
@@ -113,36 +141,36 @@ def tearDownClass(cls):
...
+    # def test_get_mgr_SourceNotFound_multi(self):
+    #     """XROOTD (RSE/PROTOCOLS): Get multiple files from storage providing LFNs and PFNs (SourceNotFound)"""
+    #     with pytest.raises(exception.SourceNotFound):
+    #         self.mtc.test_get_mgr_SourceNotFound_multi()
+    #
+    # def test_get_mgr_SourceNotFound_single_lfn(self):
+    #     """XROOTD (RSE/PROTOCOLS): Get a single file from storage providing LFN (SourceNotFound)"""
+    #     with pytest.raises(exception.SourceNotFound):
+    #         self.mtc.test_get_mgr_SourceNotFound_single_lfn()
+    #
```
Comment (maany) [inline lib/rucio/tests/test_rse_protocol_xrootd.py]: redundant changes

**Comment by maany** [comment]
## Issue 1: Most test_rse_protocol_* (except xrootd) fail

### description
In this PR, I had initially added rse_accounts.cfg inside the rucio-dev container. 
This caused all test_rse_protocol_* tests to run, locally as well.
Most of these tests failed except for test_rse_protocol_xrootd that used the xrd containers. 

### solution
I've moved the following command
```cp etc/rse_accounts.cfg.template etc/rse_accounts.cfg ```that I had initially put inside docker_activate_rses.sh to the github workflow step instead. Therefore, the test_rse_protocol_* tests get activated only in the integration tests environment on GitHub Actions. For anyone running tests locally, all the test_rse_protocol_* tests are still skipped. We should create rse_accounts.cfg available inside the rucio-dev container by default once we have fixed the protocol tests besides those for xrootd. I'll start looking into these separately.

## Issue 2: xrootd connect fails 
### Description
The execution of the command below:
https://github.com/rucio/rucio/blob/master/lib/rucio/rse/protocols/xrootd.py#L199 

via python does not work.  I dug a little deeper by running
```
cmd = '*XrdSecDEBUG=2* XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs xrd1:1094 query config xrd1:1094'
execute(cmd)
```
The output was as follows:
```
>>> subprocess.run(cmd, shell=True, stdout=subprocess.PIPE)
sec_Client: protocol request for host xrd1 token='&P=gsi,v:10400,c:ssl,ca:5fca1cb1.0|b96dc756.0'
sec_PM: Loaded gsi protocol object from libXrdSecgsi.so
201203 15:54:23 175 secgsi_Init: option CACheck: verifyss
201203 15:54:23 175 secgsi_Init: using CA dir(s): /etc/grid-security/certificates/
201203 15:54:23 175 secgsi_Init: option CRLCheck: 1 ('use-if-available'; download? no)
201203 15:54:23 175 secgsi_Init: using CRL dir(s): /etc/grid-security/certificates/
201203 15:54:23 175 secgsi_Init: CRL information refreshed every 86400 secs
201203 15:54:23 175 secgsi_Init: trust DNS option: 1
201203 15:54:23 175 secgsi_Init: using certificate file:         /opt/rucio/etc/usercert.pem
201203 15:54:23 175 secgsi_Init: using private key file:         /opt/rucio/etc/userkey.pem
201203 15:54:23 175 secgsi_Init: proxy: file:                    /tmp/x509up_u0
201203 15:54:23 175 secgsi_Init: proxy: validity:                12:00
201203 15:54:23 175 secgsi_Init: proxy: depth of signature path: 0
201203 15:54:23 175 secgsi_Init: proxy: bits in key:             512
201203 15:54:23 175 secgsi_Init: server cert: allowed names:
Secgsi -------------------------------------------------------------------
Secgsi Mode: client
Secgsi Debug: 2
Secgsi CA dir: /etc/grid-security/certificates/
Secgsi CA verification level: verifyss
Secgsi CRL dir: /etc/grid-security/certificates/
Secgsi CRL extension: .r0
Secgsi CRL check level: try
Secgsi CRL refresh time: 86400
Secgsi Certificate: /opt/rucio/etc/usercert.pem
Secgsi Key: /opt/rucio/etc/userkey.pem
Secgsi Proxy file: /tmp/x509up_u0
Secgsi Proxy validity: 12:00
Secgsi Proxy dep length: 0
Secgsi Proxy bits: 512
Secgsi Proxy sign option: 1
Secgsi Proxy delegation option: 0
Secgsi Allowed server names: [*/]<target host name>[/*]
Secgsi Crypto modules: ssl
Secgsi Ciphers: aes-128-cbc:bf-cbc:des-ede3-cbc
Secgsi MDigests: sha1:md5
Secgsi Trusting DNS for hostname checking
Secgsi -------------------------------------------------------------------
sec_PM: Using gsi protocol, args='v:10400,c:ssl,ca:5fca1cb1.0|b96dc756.0'
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: constructing: 0x7fd0300049f0
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: constructing: host: xrd1
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: p: gsi, plen: 4
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: mode: client
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: object created: v..
201203 15:54:23 175 secgsi_ParseCrypto: parsing list: ssl
201203 15:54:23 175 secgsi_ParseCrypto: found module: ssl
201203 15:54:23 175 crypto_Factory::GetCryptoFactory: shared library 'libXrdCryptossl.so' loaded
201203 15:54:23 175 sut_Rndm::GetBuffer: enter: len: 32
201203 15:54:23 175 sut_Rndm::Init: taking seed from /dev/urandom
201203 15:54:23 175 cryptossl_sslCipher::XrdCryptosslCipher: generate DH full key
201203 15:54:23 175 secgsi_ParseCAlist: parsing list: 5fca1cb1.0|b96dc756.0
201203 15:54:23 175 secgsi_GetCA: Querying cache for tag: 5fca1cb1.0:1 (timestamp:1607010863, refresh fq:86400)
201203 15:54:23 175 secgsi_GetCA: trying to load CA certificate from /etc/grid-security/certificates/5fca1cb1.0
201203 15:54:23 175 cryptossl_X509::CertType: certificate has 3 extensions
201203 15:54:23 175 cryptossl_X509::CertType: CA certificate
201203 15:54:23 175 cryptossl_X509ParseFile: certificate for '/CN=Rucio Development CA'added to the chain - ord: 1
201203 15:54:23 175 cryptossl_X509ParseFile: no RSA private key found in file /etc/grid-security/certificates/5fca1cb1.0
201203 15:54:23 175 cryptossl_ASN1toUTC:  UTC: 1565374203  isdst: 0
201203 15:54:23 175 cryptossl_ASN1toUTC:  UTC: 2429287803  isdst: 0
201203 15:54:23 175 secgsi_LoadCRL: target file: /etc/grid-security/certificates/5fca1cb1.r0
201203 15:54:23 175 cryptossl_X509Crl::Init: file /etc/grid-security/certificates/5fca1cb1.r0 does not exist - do nothing
201203 15:54:23 175 cryptossl_X509Crl::XrdCryptosslX509Crl_file: could not initialize the CRL from /etc/grid-security/certificates/5fca1cb1.r0
201203 15:54:23 175 secgsi_GetCA: CRL is missing or expired: ignoring (CRLCheck: 1)
201203 15:54:23 175 cryptossl_X509ParseFile: unable to open file (errno: 2)
201203 15:54:23 175 secgsi_QueryProxy: proxy files must have at least two certificates (found: 0)
201203 15:54:23 175 secgsi_InitProxy: Not a tty: cannot prompt for proxies - do nothing
201203 15:54:23 175 secgsi_QueryProxy: problems initializing proxy via external shell
201203 15:54:23 175 secgsi_getCredentials: error getting user proxies CF: 0x7fd02ffffbc0
201203 15:54:23 175 secgsi_ErrF: Secgsi: ErrParseBuffer: error getting user proxies: kXGS_init
sec_Client: protocol request for host xrd1 token=' '
[FATAL] Auth failed
```

The line to be concerned about is:
```
201203 15:54:23 175 secgsi_QueryProxy: problems initializing proxy via external shell
```

Which makes sense since we are executing this command in a subshell:
https://github.com/rucio/rucio/blob/28bf965377a1271d25caf1424745dd8b99eef12f/lib/rucio/common/utils.py#L463

As a result, the tests get RseAccessDenied error

### Proposed Solution
Currently, I execute the command here (not in an external shell):
https://github.com/rucio/rucio/pull/4165/files#diff-0e09f3b5aa912ee500a40d52d7cd4e628cbc14c491d78fdeba858902fbf3f1a7R35-R37
which fixes the issue. 

## Issue 3: Hardcoded container rse info
### Description
This one is minor. I search for containers in the testing environment. If found, the rse_info is set to reflect the container's rse
https://github.com/rucio/rucio/pull/4165/files#diff-d18a713b2ba7a8059c56d7ee66533919b152cbc83c2bd9567826ad17f6f46bafR67-R70

Currently, the prefix, hostname, port are hardcoded. 

### Proposed Solution
I've made a note of this. I plan to change this code in a separate PR when I have more protocol tests ready. The goal is to have a consistent way of detecting containerized test environments and picking up rse info accordingly. I want to see if a consistent approch would be possible for multiple protocols.

File: .github/workflows/integration_tests.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  - pull_request
+  - push
+  - workflow_dispatch
+
+jobs:
+  integration-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Start containers
```
Comment (bziemons) [inline .github/workflows/integration_tests.yml:66]: I think we need to build the rucio dev container from source before starting the cluster or otherwise tests may fail when a new API endpoint is used or when rucio dependencies change.

Comment (maany) [reply .github/workflows/integration_tests.yml:66]: @bziemons I've modified the workflow as follows:
- build from src + use any cached image layers
- uploading image/cached layers to GitHub registry
- run the tests.

It works in my fork:
https://github.com/maany/rucio/runs/1492936906?check_suite_focus=true 

The access to rucio's git registry seems to be restricted for PRs. 
https://github.com/rucio/rucio/pull/4165/checks?check_run_id=1492937164#step:6:1739 

It's an ongoing issue for GitHub Actions:
https://github.com/actions/first-interaction/issues/10 

It should be fixed when the PR is merged IMO.

**Comment by maany** [comment]
Ready to be merged!
Explanation of the failing CI check follows

I've modified the workflow as follows:

- build from src + use any cached image layers
- uploading image/cached layers to GitHub registry
- run the tests.

It works in my fork:
https://github.com/maany/rucio/runs/1492936906?check_suite_focus=true

The access to rucio's git registry seems to be restricted for PRs, since github.actor is @maany .
https://github.com/rucio/rucio/pull/4165/checks?check_run_id=1492937164#step:6:1739

It's an ongoing issue for GitHub Actions:
actions/first-interaction#10

In our case, the integration tests should pass when the PR is merged.

File: .github/workflows/integration_tests.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      - uses: actions/checkout@v2
+        name: Checkout rucio source
+        with:
+          path: dev/rucio
+      - name: Prepare Images
+        shell: bash
+        run: |
+          sed -i "s/RUN git clone .*/COPY .\/rucio \/tmp\/rucio/" $GITHUB_WORKSPACE/dev/Dockerfile_py3
+      - name: Build Images
+        uses: whoan/docker-build-with-cache-action@v5
```
Comment (bziemons) [inline .github/workflows/integration_tests.yml]: as a general question: is it fine to trust third party code with the GITHUB_TOKEN ? Malicious code could edit issues/PRs/source code on the repo. So far I only trusted the GitHub curated actions. CC: @bari12

File: .github/workflows/integration_tests.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          repository: rucio/containers
+      - uses: actions/checkout@v2
+        name: Checkout rucio source
+        with:
+          path: dev/rucio
+      - name: Prepare Images
+        shell: bash
+        run: |
+          sed -i "s/RUN git clone .*/COPY .\/rucio \/tmp\/rucio/" $GITHUB_WORKSPACE/dev/Dockerfile_py3
+      - name: Build Images
```
Comment (bziemons) [inline .github/workflows/integration_tests.yml]: Moving the build and push (of the container images) to the `imagecache` workflow and instead just pulling and re-building here would resolve the issue with the PR not being able to run this testsuite.
Since the run is currently failing here, I am assuming that it will fail with any new PR, which makes this testsuite less useful. Without pushing the image to the GitHub image repository in this workflow it should work fine like the existing test workflows.

File: .github/workflows/integration_tests.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+          sed "s/image: rucio\/rucio-dev/image: $IMG/" $GITHUB_WORKSPACE/dev/rucio/etc/docker/dev/docker-compose-storage.yml
+      - name: Start containers
+        run: docker-compose -f $GITHUB_WORKSPACE/dev/rucio/etc/docker/dev/docker-compose-storage.yml up -d
+      - name: Initialize tests
+        shell: bash
+        run: |
+          docker exec -t dev_rucio_1 cp etc/rse-accounts.cfg.template etc/rse-accounts.cfg
+          docker exec -t dev_rucio_1 tools/run_tests_docker.sh -ir
+      - name: File Upload/Download Test
+        run: docker exec -t dev_rucio_1 pytest -v --full-trace lib/rucio/tests/test_rucio_server.py
```
Comment (bziemons) [inline .github/workflows/integration_tests.yml]: there was a slight change in how the tests are run now: `python -bb -m pytest -vvvrxs …`. I will definitely replace this cmdline with a documented helper tool soon, so that can be reused instead of this. So this change is optional on your side.

File: lib/rucio/tests/test_rse_protocol_xrootd.py [hunk 7]
```diff
@@ -113,36 +141,36 @@ def tearDownClass(cls):
...
+    # def test_get_mgr_SourceNotFound_single_pfn(self):
+    #     """XROOTD (RSE/PROTOCOLS): Get a single file from storage providing PFN (SourceNotFound)"""
+    #     with pytest.raises(exception.SourceNotFound):
+    #         self.mtc.test_get_mgr_SourceNotFound_single_pfn()
```
Comment (bziemons) [inline lib/rucio/tests/test_rse_protocol_xrootd.py]: I would prefer removal over commenting out chunks of code :)

**Review by bziemons — CHANGES_REQUESTED** [review]
This PR looks generally good. Just a few things left to do.

File: .github/workflows/integration_tests.yml (hunk 4 above)
Comment (mlassnig) [reply .github/workflows/integration_tests.yml]: I'm not particularly comfortable with this. Is there a safer way to do this?

File: .github/workflows/integration_tests.yml (hunk 6 above)
Comment (maany) [reply .github/workflows/integration_tests.yml]: done :)

File: .github/workflows/integration_tests.yml (hunk 4 above)
Comment (maany) [reply .github/workflows/integration_tests.yml]: Sure, I've removed it

File: .github/workflows/autotest.yml [hunk 8]
```diff
@@ -15,7 +15,7 @@ jobs:
         run: python3 -m pip install -U PyYAML
       - name: Identify Matrix
         id: matrix
-        run: echo "::set-output name=matrix::$(./tools/test/matrix_parser.py < ./etc/docker/test/matrix.yml)"
+        run: echo "::set-output name=matrix::$(jq -c '[.[]  | select(.SUITE!="integration-test")]' <<< $(./tools/test/matrix_parser.py < ./etc/docker/test/matrix.yml))"
```
Comment (bziemons) [inline .github/workflows/autotest.yml]: I was thinking about renaming + adding a second matrix file if something like this was ever needed :) However this is +1 on creativity

File: .github/workflows/integration_tests.yml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            fi
+          done
+          docker logout https://docker.pkg.github.com
+          if [[ -z $IMAGES ]]; then exit 1; fi
+          echo "::set-output name=images::$IMAGES"
+      - name: Prepare Docker Compose
+        shell: bash
+        run: |
+          IMG=$(docker image ls -q --filter "reference=docker.pkg.github.com/${{ github.repository }}/rucio-integration-test")
+          sed "s/image: rucio\/rucio-dev.*/image: $IMG/" $GITHUB_WORKSPACE/etc/docker/dev/docker-compose-storage.yml
```
Comment (bziemons) [inline .github/workflows/integration_tests.yml]: there is no need to de-reference the image tag:
```sh
sed -i "s;image: rucio/rucio-dev.*;image: docker.pkg.github.com/${{ github.repository }}/rucio-integration-test;" …
```

Also you are missing the `-i` on the sed, the integration tests currently do not use the previously built images:
`Status: Downloaded newer image for rucio/rucio-dev:py3`

File: tools/test/build_images.py [hunk 10]
```diff
@@ -91,6 +75,38 @@ def main():
...
+    script_args = parser.parse_args()
+
+    integration_tests, autotests = [], []
+    for x in matrix:
+        autotests.append(x) if x['SUITE'] != 'integration-test' else integration_tests.append(x)
+
+    images = dict()
+
+    autotest_images = build_images(matrix, script_args, test_mode='autotest')
+    integration_test_images = build_images(matrix, script_args, test_mode='integration-test')
```
Comment (bziemons) [inline tools/test/build_images.py]: there is no check here if the matrix' SUITE contains only integration tests or not, so both images are always built - see the GH Actions log on this PR. I would rather have this `test_mode` directly in the matrix configured and renamed to `image_identifier` which defaults to `autotest`.
The built images do base on the `centos7.Dockerfile` in this repo instead of the `rucio-dev` image from the containers repo. Not sure if that will be a problem, since they differ.

File: tools/test/build_images.py [hunk 11]
```diff
@@ -91,6 +75,38 @@ def main():
...
+                        help='build images without cache')
+    parser.add_argument('-r', '--cache-repo', dest='cache_repo', type=str, default='docker.pkg.github.com/rucio/rucio',
+                        help='use the following cache repository, like docker.pkg.github.com/USER/REPO')
+    parser.add_argument('-p', '--push-cache', dest='push_cache', action='store_true',
+                        help='push the images to the cache repo')
+    script_args = parser.parse_args()
+
+    integration_tests, autotests = [], []
+    for x in matrix:
+        autotests.append(x) if x['SUITE'] != 'integration-test' else integration_tests.append(x)
```
Comment (bziemons) [inline tools/test/build_images.py]: dead code?

**Review by bziemons — CHANGES_REQUESTED** [review]

File: .github/workflows/integration_tests.yml (hunk 9 above)
Comment (maany) [reply .github/workflows/integration_tests.yml]: @bziemons good catch! I Will be more careful. Thanks!

File: tools/test/run_tests.py [hunk 12]
```diff
@@ -83,6 +84,8 @@ def case_log(caseid, msg, file=sys.stderr):
 
 def main():
     obj = json.load(sys.stdin)
+    print(obj)
+    print("*********")
```
Comment (bziemons) [inline tools/test/run_tests.py]: remove debug output, revert this file

File: .github/workflows/integration_tests.yml (hunk 9 above)
Comment (bziemons) [reply .github/workflows/integration_tests.yml]: I would still like the de-referencing of the image tag gone. It adds unneeded complication and sources of errors

File: .github/workflows/autotest.yml [hunk 13]
```diff
@@ -44,6 +46,9 @@ jobs:
           docker logout https://docker.pkg.github.com
           if [[ -z $IMAGES ]]; then exit 1; fi
           echo "::set-output name=images::$IMAGES"
+      - name: Test output
+        run: |
+          echo '{"matrix": ${{ toJson(matrix.cfg) }}, "images": ${{ steps.images.outputs.images }} }'
```
Comment (bziemons) [inline .github/workflows/autotest.yml]: this debug output is already in the GH Actions logs, please remove these two (in `autotest.yml`)

**Review by bziemons — CHANGES_REQUESTED** [review]

**Comment by bziemons** [comment]
The rest of GH Actions workflow changes look fine. We can disregard the `release-patch-setup` since @bari12 might not want to merge this back to release branches anyway. Also we usually squash our changes into one commit.

**Comment by bziemons** [comment]
@bari12 @mlassnig can somebody have a look over the changes in `lib/rucio/tests/rsemgr_api_test.py` and `lib/rucio/tests/test_rse_protocol_xrootd.py` ? I can't fully judge them.

**Review by bziemons — APPROVED** [review]

**Review by bari12 — COMMENTED** [review]
Hi @maany 
Apologies for the massive delay of comments.
To me it looks good. Also the changes in the test files. I am still wondering a bit if using the dev container is the best approach, due to it's potential volatility, but creating a new container will be yet another thing to manage, so also not sure if this is better.

The tests look fine, the only thing I noticed is though: We eventually want to remove the `rsemgr.upload` functionality. This is largely deprecated in favour of the `uploadclient` now. I know that the tests are in rsemgr_api_test, so it should be about the rsemgr, but maybe in parallel we should do this as well with the uploadclient.

👍 Thanks a lot.

File: lib/rucio/rse/protocols/xrootd.py [hunk 14]
```diff
@@ -114,16 +114,14 @@ def stat(self, path):
             self.logger.info('xrootd.stat: filesize cmd: {}'.format(cmd))
             status_stat, out, err = execute(cmd)
             if status_stat == 0:
-                ret['filesize'] = out.split(b'\n')[2].split()[-1].decode()
+                ret['filesize'] = out.split('\n')[2].split()[-1]
```
Comment (mlassnig) [inline lib/rucio/rse/protocols/xrootd.py:117]: why did you change this here? i remember adding this explicitly to keep compatible with some clients

File: tools/docker_activate_rses.sh [hunk 15]
```diff
@@ -22,9 +22,19 @@ rucio-admin rse add XRD2
...
+rucio-admin rse add-protocol --hostname xrd2 --scheme root --prefix //rucio --port 1095 --impl rucio.rse.protocols.xrootd.Default --domain-json '{"wan": {"read": 1, "write": 1, "delete": 1, "third_party_copy": 1}, "lan": {"read": 1, "write": 1, "delete": 1}}' XRD2
+rucio-admin rse add-protocol --hostname xrd3 --scheme root --prefix //rucio --port 1096 --impl rucio.rse.protocols.xrootd.Default --domain-json '{"wan": {"read": 1, "write": 1, "delete": 1, "third_party_copy": 1}, "lan": {"read": 1, "write": 1, "delete": 1}}' XRD3
+
+# Set test_container_xrd attribute for xrd containers
+rucio-admin rse set-attribute --rse XRD1 --key test_container_xrd --value True
+rucio-admin rse set-attribute --rse XRD2 --key test_container_xrd --value True
+rucio-admin rse set-attribute --rse XRD3 --key test_container_xrd --value True
+
+# Workaround, xrootd.py#connect returns with Auth Failed due to execution of the command in subprocess
+XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs xrd1:1094 query config xrd1:1094
```
Comment (mlassnig) [inline tools/docker_activate_rses.sh:35]: you don't do anything with the return values here?

File: tools/test/run_tests.py [hunk 16]
```diff
@@ -68,6 +68,7 @@ def find_image(images: typing.Dict, case: typing.Dict):
 
 def case_id(case: typing.Dict) -> str:
     parts = [
+
```
Comment (mlassnig) [inline tools/test/run_tests.py:71]: extra blank line?

File: lib/rucio/rse/protocols/xrootd.py (hunk 14 above)
Comment (maany) [reply lib/rucio/rse/protocols/xrootd.py:117]: Hey @mlassnig,
When I run the test_rse_protocol_xrootd tests (in dev/[Dockerfile_py3](https://github.com/rucio/containers/blob/master/dev/Dockerfile_py3) container) with 

```
ret['filesize'] = out.split(b'\n')[2].split()[-1].decode()
```

The following tests fail:
```
lib/rucio/tests/test_rse_protocol_xrootd.py::TestRseXROOTD::test_put_mgr_ok_multi FAILED                                                                                                             [ 62%]
lib/rucio/tests/test_rse_protocol_xrootd.py::TestRseXROOTD::test_put_mgr_ok_single FAILED
```

The stack trace is as follows:
```
            status_stat, out, err = execute(cmd)
            if status_stat == 0:
>               ret['filesize'] = out.split(b'\n')[2].split()[-1].decode()
E               TypeError: must be str or None, not bytes

lib/rucio/rse/protocols/xrootd.py:117: TypeError

```
Therefore, I had changed the separator type to `str`

File: tools/docker_activate_rses.sh (hunk 15 above)
Comment (maany) [reply tools/docker_activate_rses.sh:35]: @mlassnig 
I can add:
```
XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs xrd1:1094 query config xrd1:1094 && echo "xrd1 ok" || echo "RSEAccessDenied: xrd1"
```

I've explained a bit above on why I've added this workaround here. I'm copying that text below:
 
## Issue 2: xrootd connect fails 
### Description
The execution of the command below:
https://github.com/rucio/rucio/blob/master/lib/rucio/rse/protocols/xrootd.py#L199 

via python does not work.  I dug a little deeper by running
```
cmd = '*XrdSecDEBUG=2* XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs xrd1:1094 query config xrd1:1094'
execute(cmd)
```
The output was as follows:
```
>>> subprocess.run(cmd, shell=True, stdout=subprocess.PIPE)
sec_Client: protocol request for host xrd1 token='&P=gsi,v:10400,c:ssl,ca:5fca1cb1.0|b96dc756.0'
sec_PM: Loaded gsi protocol object from libXrdSecgsi.so
201203 15:54:23 175 secgsi_Init: option CACheck: verifyss
201203 15:54:23 175 secgsi_Init: using CA dir(s): /etc/grid-security/certificates/
201203 15:54:23 175 secgsi_Init: option CRLCheck: 1 ('use-if-available'; download? no)
201203 15:54:23 175 secgsi_Init: using CRL dir(s): /etc/grid-security/certificates/
201203 15:54:23 175 secgsi_Init: CRL information refreshed every 86400 secs
201203 15:54:23 175 secgsi_Init: trust DNS option: 1
201203 15:54:23 175 secgsi_Init: using certificate file:         /opt/rucio/etc/usercert.pem
201203 15:54:23 175 secgsi_Init: using private key file:         /opt/rucio/etc/userkey.pem
201203 15:54:23 175 secgsi_Init: proxy: file:                    /tmp/x509up_u0
201203 15:54:23 175 secgsi_Init: proxy: validity:                12:00
201203 15:54:23 175 secgsi_Init: proxy: depth of signature path: 0
201203 15:54:23 175 secgsi_Init: proxy: bits in key:             512
201203 15:54:23 175 secgsi_Init: server cert: allowed names:
Secgsi -------------------------------------------------------------------
Secgsi Mode: client
Secgsi Debug: 2
Secgsi CA dir: /etc/grid-security/certificates/
Secgsi CA verification level: verifyss
Secgsi CRL dir: /etc/grid-security/certificates/
Secgsi CRL extension: .r0
Secgsi CRL check level: try
Secgsi CRL refresh time: 86400
Secgsi Certificate: /opt/rucio/etc/usercert.pem
Secgsi Key: /opt/rucio/etc/userkey.pem
Secgsi Proxy file: /tmp/x509up_u0
Secgsi Proxy validity: 12:00
Secgsi Proxy dep length: 0
Secgsi Proxy bits: 512
Secgsi Proxy sign option: 1
Secgsi Proxy delegation option: 0
Secgsi Allowed server names: [*/]<target host name>[/*]
Secgsi Crypto modules: ssl
Secgsi Ciphers: aes-128-cbc:bf-cbc:des-ede3-cbc
Secgsi MDigests: sha1:md5
Secgsi Trusting DNS for hostname checking
Secgsi -------------------------------------------------------------------
sec_PM: Using gsi protocol, args='v:10400,c:ssl,ca:5fca1cb1.0|b96dc756.0'
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: constructing: 0x7fd0300049f0
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: constructing: host: xrd1
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: p: gsi, plen: 4
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: mode: client
201203 15:54:23 175 secgsi_XrdSecProtocolgsi: object created: v..
201203 15:54:23 175 secgsi_ParseCrypto: parsing list: ssl
201203 15:54:23 175 secgsi_ParseCrypto: found module: ssl
201203 15:54:23 175 crypto_Factory::GetCryptoFactory: shared library 'libXrdCryptossl.so' loaded
201203 15:54:23 175 sut_Rndm::GetBuffer: enter: len: 32
201203 15:54:23 175 sut_Rndm::Init: taking seed from /dev/urandom
201203 15:54:23 175 cryptossl_sslCipher::XrdCryptosslCipher: generate DH full key
201203 15:54:23 175 secgsi_ParseCAlist: parsing list: 5fca1cb1.0|b96dc756.0
201203 15:54:23 175 secgsi_GetCA: Querying cache for tag: 5fca1cb1.0:1 (timestamp:1607010863, refresh fq:86400)
201203 15:54:23 175 secgsi_GetCA: trying to load CA certificate from /etc/grid-security/certificates/5fca1cb1.0
201203 15:54:23 175 cryptossl_X509::CertType: certificate has 3 extensions
201203 15:54:23 175 cryptossl_X509::CertType: CA certificate
201203 15:54:23 175 cryptossl_X509ParseFile: certificate for '/CN=Rucio Development CA'added to the chain - ord: 1
201203 15:54:23 175 cryptossl_X509ParseFile: no RSA private key found in file /etc/grid-security/certificates/5fca1cb1.0
201203 15:54:23 175 cryptossl_ASN1toUTC:  UTC: 1565374203  isdst: 0
201203 15:54:23 175 cryptossl_ASN1toUTC:  UTC: 2429287803  isdst: 0
201203 15:54:23 175 secgsi_LoadCRL: target file: /etc/grid-security/certificates/5fca1cb1.r0
201203 15:54:23 175 cryptossl_X509Crl::Init: file /etc/grid-security/certificates/5fca1cb1.r0 does not exist - do nothing
201203 15:54:23 175 cryptossl_X509Crl::XrdCryptosslX509Crl_file: could not initialize the CRL from /etc/grid-security/certificates/5fca1cb1.r0
201203 15:54:23 175 secgsi_GetCA: CRL is missing or expired: ignoring (CRLCheck: 1)
201203 15:54:23 175 cryptossl_X509ParseFile: unable to open file (errno: 2)
201203 15:54:23 175 secgsi_QueryProxy: proxy files must have at least two certificates (found: 0)
201203 15:54:23 175 secgsi_InitProxy: Not a tty: cannot prompt for proxies - do nothing
201203 15:54:23 175 secgsi_QueryProxy: problems initializing proxy via external shell
201203 15:54:23 175 secgsi_getCredentials: error getting user proxies CF: 0x7fd02ffffbc0
201203 15:54:23 175 secgsi_ErrF: Secgsi: ErrParseBuffer: error getting user proxies: kXGS_init
sec_Client: protocol request for host xrd1 token=' '
[FATAL] Auth failed
```

The line to be concerned about is:
```
201203 15:54:23 175 secgsi_QueryProxy: problems initializing proxy via external shell
```

Which makes sense since we are executing this command in a subshell:
https://github.com/rucio/rucio/blob/28bf965377a1271d25caf1424745dd8b99eef12f/lib/rucio/common/utils.py#L463

As a result, the tests get RseAccessDenied error

### Proposed Solution
Currently, I execute the command here (not in an external shell):
https://github.com/rucio/rucio/pull/4165/files#diff-0e09f3b5aa912ee500a40d52d7cd4e628cbc14c491d78fdeba858902fbf3f1a7R35-R37
which fixes the issue.

**Comment by bziemons** [comment]
The py2 tests need this fix: https://github.com/rucio/rucio/pull/4267 which is on master already. This can still be merged as is.
